### PR TITLE
Fix ES6 imports of CommonJs modules

### PR DESCRIPTION
**Pull request by billti**
Changes to address issue #7076.  This is not yet done, but wanted to get feedback on the approach before continuing further from @vladima, @mhegazy, and/or @RyanCavanaugh .

Essentially the way the `export=` symbol for `module.exports = ...` in CommonJS files was being created was quite different to the `export=` in TypeScript, even though the same symbol in `sourceFile.symbol.exports` is used for both. This was not a problem for `require(..)` calls, where the type checking of the `require` call expression was special cased. However for imports using ES2015 syntax, this was failing as it expects that a) The `export=` symbol is an alias, and b) It is an alias to a module or an identifier.

This code starts to address some of this, but a few questions remain.

 - I've hacked in special handling for `module.export =` where it's an object literal or an identifier. This may well be the wrong approach, and perhaps should just synthesize a symbol/type for the expression being assigned. For example, `module.exports = getSomeObj()` is perfectly valid.
 - ES2015 imports have different semantics/restrictions, so should we honor those (i.e. the imported alias can't have call/construct signatures - which we'd have to remove).

Note that I've built this branch on top of the "load JS from node_modules" work, as there was some tests there that covered this (it was doing that work I undercovered this issue). It's really only the last commit ( bf030a2 ) which is of interest for this fix.

Let's discuss tomorrow on the right approach before I spend more time on this...

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1005,6 +1005,13 @@ namespace ts {
                     return getTargetOfExportSpecifier(<ExportSpecifier>node);
                 case SyntaxKind.ExportAssignment:
                     return getTargetOfExportAssignment(<ExportAssignment>node);
+                case SyntaxKind.BinaryExpression:
+                    // module.exports = ...
+                    const targetExpr = (node as BinaryExpression).right;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: wrap this in a helper to match the style of the other conditions,

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -1389,8 +1389,13 @@ namespace ts {
                 declareSymbol(container.symbol.exports, container.symbol, node, SymbolFlags.Alias, SymbolFlags.PropertyExcludes | SymbolFlags.AliasExcludes);
             }
             else {
+                if (getSpecialPropertyAssignmentKind(node) === SpecialPropertyAssignmentKind.ModuleExports) {
+                    declareSymbol(container.symbol.exports, container.symbol, node, SymbolFlags.Alias, SymbolFlags.PropertyExcludes | SymbolFlags.AliasExcludes);
```
Comment (mhegazy) [inline src/compiler/binder.ts]: we should check that the right hand side is an identifier

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -1005,6 +1005,13 @@ namespace ts {
                     return getTargetOfExportSpecifier(<ExportSpecifier>node);
                 case SyntaxKind.ExportAssignment:
                     return getTargetOfExportAssignment(<ExportAssignment>node);
+                case SyntaxKind.BinaryExpression:
+                    // module.exports = ...
+                    const targetExpr = (node as BinaryExpression).right;
+                    if (targetExpr.kind === SyntaxKind.Identifier) {
+                        return resolveName(node, (targetExpr as Identifier).text, SymbolFlags.Value, Diagnostics.Cannot_find_name_0, (targetExpr as Identifier));
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why not Namespace as well, how would `import {x} from 'mod'" work then?

**Comment by DanielRosenwasser** [comment]
> Note that I've built this branch on top of the "load JS from node_modules" work, as there was some tests there that covered this

You can interactively rebase the branch and filter out the irrelevant commits to fix this up.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -1005,6 +1005,13 @@ namespace ts {
                     return getTargetOfExportSpecifier(<ExportSpecifier>node);
                 case SyntaxKind.ExportAssignment:
                     return getTargetOfExportAssignment(<ExportAssignment>node);
+                case SyntaxKind.BinaryExpression:
+                    // module.exports = ...
+                    const targetExpr = (node as BinaryExpression).right;
+                    if (targetExpr.kind === SyntaxKind.Identifier) {
+                        return resolveName(node, (targetExpr as Identifier).text, SymbolFlags.Value, Diagnostics.Cannot_find_name_0, (targetExpr as Identifier));
+                    }
+                    return getSymbolOfNode((node as BinaryExpression).right);
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: `getSymbolOfNode(targetExpr)`

**Comment by billti** [comment]
Closing in favor of #7249.

**Pull request closed without merging**